### PR TITLE
blk/zoned: Add reset_zones functionality for zoned devices

### PR DESCRIPTION
**Pull request by Rishabh4275**
Signed-off-by: Rishabh Chawla <rishabhchawla1995@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Rishabh4275** [comment]
Values after reset
Values before reset

**Comment by Rishabh4275** [comment]
Values before reset 
Values before reset
Values after reset 
Values after reset

File: src/blk/zoned/HMSMRDevice.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ class HMSMRDevice final : public BlockDevice {
   string vdo_name;
 
   std::string devname;  ///< kernel dev name (/sys/block/$devname), if any
+  int zbd_file_descriptor; /// zbd_open File Descriptor
```
Comment (agayev) [inline src/blk/zoned/HMSMRDevice.h:44]: I'd say call this just zbd_dev.

File: src/blk/zoned/HMSMRDevice.cc [hunk 2]
```diff
@@ -412,6 +412,14 @@ void HMSMRDevice::_detect_vdo()
   return;
 }
 
+// Will reset the write pointer of all zones from start to end(both inclusive)
+bool HMSMRDevice::reset_zones(uint64_t zone_num_range_start, uint64_t zone_num_range_end) {
```
Comment (agayev) [inline src/blk/zoned/HMSMRDevice.cc:416]: Return int to be consistent with the remaining functions.

Nit: Omit range from parameter names.

File: src/blk/zoned/HMSMRDevice.cc [hunk 3]
```diff
@@ -412,6 +412,14 @@ void HMSMRDevice::_detect_vdo()
...
+  ceph_assert(is_smr());
+  uint64_t len = (zone_num_range_end + 1 - zone_num_range_start)  * zone_size;
+  ceph_assert(len > 0);
+  return !zbd_reset_zones(zbd_file_descriptor, zone_num_range_start * zone_size, len);
```
Comment (agayev) [inline src/blk/zoned/HMSMRDevice.cc:420]: In case of failure it would be useful to print the error message.

File: src/blk/BlockDevice.h [hunk 4]
```diff
@@ -200,6 +200,11 @@ class BlockDevice {
     return conventional_region_size;
   }
 
+  virtual bool reset_zones(uint64_t zone_num_range_start, uint64_t zone_num_range_end) {
+    ceph_assert(is_smr());
```
Comment (agayev) [inline src/blk/BlockDevice.h:204]: Not sure if we should assert here given that right below we return false. Let's see what Ceph folks say.

**Comment by agayev** [comment]
@tchaikov not sure if @ifed01 or someone else who is responsible forl blk should review this. Please tag the appropriate person. Thanks!

File: src/blk/zoned/HMSMRDevice.h (hunk 1 above)
Comment (Rishabh4275) [reply src/blk/zoned/HMSMRDevice.h:44]: Updated

File: src/blk/zoned/HMSMRDevice.cc (hunk 2 above)
Comment (Rishabh4275) [reply src/blk/zoned/HMSMRDevice.cc:416]: Updated.
Thank you.

File: src/blk/zoned/HMSMRDevice.cc (hunk 3 above)
Comment (Rishabh4275) [reply src/blk/zoned/HMSMRDevice.cc:420]: Added

**Comment by agayev** [comment]
@tchaikov can you please assign this to a reviewer? Thanks.

File: src/blk/BlockDevice.h [hunk 5]
```diff
@@ -200,6 +200,11 @@ class BlockDevice {
     return conventional_region_size;
   }
 
+  virtual uint64_t reset_zones(uint64_t zone_num_start, uint64_t zone_num_end) {
```
Comment (tchaikov) [inline src/blk/BlockDevice.h:203]: is this method called anywhere?

Comment (agayev) [reply src/blk/BlockDevice.h:203]: Not yet. It will be called in the next PR by the code that cleans zones.

Comment (tchaikov) [reply src/blk/BlockDevice.h:203]: i'd like to see how it is used for better understanding of this change. IMHO, it'd be better to have the change which introduces the consumer in the same PR.

Comment (agayev) [reply src/blk/BlockDevice.h:203]: Okay, we will do that.

**Comment by Rishabh4275** [comment]
Shifting to https://github.com/ceph/ceph/pull/40923

**Pull request closed without merging**